### DISCUSSION TITLE
Testing setup with `push()`

### DISCUSSION THREAD

**Discussion by MartinP7r** [hmlongco/Factory#74] [Q&A] [answered]
Hello @hmlongco 👋🏼 

I have trouble understanding the purpose of `push()` here

https://github.com/hmlongco/Factory/blob/c0b4e777a1ccc3e53005f203c3cd0dbd949f8c17/Sources/Factory/Factory.docc/Debugging/Testing.md?plain=1#L18-L22

Could you give some explanation as to why it's necessary?  
Also, it seems that this method is on the manager class, so need to do

```swift
Container.shared.manager.push()
```
unless I'm misunderstanding something?

Cheers,
Martin

**Comment by hmlongco** [comment, marked as answer]
You're right about the manager. Still not catching everything changed from old documentation.

Push saves all of the current registrations and scope caches onto a stack. You can then do whatever registrations and changes are needed for a given test, then in the teardown you pop the stack, eliminating all of your changes and restoring the container to its original state before the push.

This lets each set of tests start from the same state, irregardless of whatever the prior test had changed.
